hwy/tests/swizzle_test.cc

```diff
@@ -343,6 +343,154 @@ HWY_NOINLINE void TestAllTwoTablesLookupLanes() {
+    if (N < 4) {
+      const RebindToSigned<decltype(d)> di;
+      using TI = TFromD<decltype(di)>;
+      const auto lane_0_valid =
+          Set(di, static_cast<TI>(-static_cast<int>(idx0 < N)));
+      if (N > 1) {
+        const auto lane_1_valid =
+            Set(di, static_cast<TI>(-static_cast<int>(idx1 < N)));
+        return RebindMask(d, MaskFromVec(InterleaveMaskVectors(di, lane_0_valid,
+                                                               lane_1_valid)));
+      }
+      return RebindMask(d, MaskFromVec(lane_0_valid));
+    }
+
+    return FirstN(d, N);
+  }
+
+  template <size_t kIdx0, size_t kIdx1, size_t kIdx2, size_t kIdx3, class D>
+  static HWY_INLINE void DoTestPer4LaneBlkShuf(
+      D d, const size_t N, const VFromD<D> v,
+      const TFromD<D>* HWY_RESTRICT src_lanes,
+      TFromD<D>* HWY_RESTRICT expected) {
+    for (size_t i = 0; i < N; i += 4) {
+      expected[i] = src_lanes[i + kIdx0];
+      expected[i + 1] = src_lanes[i + kIdx1];
+      expected[i + 2] = src_lanes[i + kIdx2];
+      expected[i + 3] = src_lanes[i + kIdx3];
+    }
+
+    if (N < 4) {
+      if (kIdx0 >= N) expected[0] = TFromD<D>{0};
+      if (kIdx1 >= N) expected[1] = TFromD<D>{0};
+    }
+
+    const auto actual = Per4LaneBlockShuffle<kIdx0, kIdx1, kIdx2, kIdx3>(v);
+    const auto valid_mask = Per4LaneBlockShufValidMask(d, N, kIdx0, kIdx1);
+    HWY_ASSERT_VEC_EQ(d, expected, IfThenElseZero(valid_mask, actual));
+
+    if (N < 4) {
+      if (kIdx0 >= N) expected[0] = src_lanes[0];
+      if (kIdx1 >= N) expected[1] = src_lanes[1];
+      HWY_ASSERT_VEC_EQ(d, expected, IfThenElse(valid_mask, actual, v));
+    }
+  }
+
+  template <class D>
+  static HWY_INLINE Vec<D> GenerateTestVect(hwy::NonFloatTag /*tag*/, D d) {
+    const RebindToUnsigned<decltype(d)> du;
+    using TU = TFromD<decltype(du)>;
+    constexpr TU kIotaStart = static_cast<TU>(0x0706050403020101u);
+    return BitCast(d, Iota(du, kIotaStart));
+  }
+
+  template <class D>
+  static HWY_INLINE Vec<D> GenerateTestVect(hwy::FloatTag /*tag*/, D d) {
+    const RebindToUnsigned<decltype(d)> du;
+    using T = TFromD<decltype(d)>;
+    using TU = TFromD<decltype(du)>;
+
+    constexpr size_t kNumOfBitsInT = sizeof(T) * 8;
+    constexpr TU kIntBitsMask =
+        (kNumOfBitsInT > 16) ? static_cast<TU>(static_cast<TU>(~TU{0}) >> 16)
+                             : TU{0};
+
+    const auto flt_iota = Set(d, 1);
+    if (kIntBitsMask == 0) return flt_iota;
+
+    const auto int_iota =
+        And(GenerateTestVect(hwy::NonFloatTag(), du), Set(du, kIntBitsMask));
+    return Or(flt_iota, BitCast(d, int_iota));
+  }
+
+  template <class D>
+  static HWY_INLINE void DoTestPer4LaneBlkShuffles(
+      D d, const size_t N, const VFromD<D> v, TFromD<D>* HWY_RESTRICT src_lanes,
+      TFromD<D>* HWY_RESTRICT expected) {
+    Store(v, d, src_lanes);
+    DoTestPer4LaneBlkShuf<0, 0, 0, 0>(d, N, v, src_lanes, expected);
```

Comment:
How about we make this truly exhaustive for safety? To avoid 256 lines, we can have a function ForeachIdx0 that calls ForeachIdx1<i> for i=0..3 which then calls ForeachIdx2<i, j> etc.

Reply:
The issue that I was running to with testing all 256 cases of `Per4LaneBlockShuffle<kIdx3, kIdx2, kIdx1, kIdx0>(v)` using the `ForeachIdx1<i>`, `ForeachIdx2<i, j>`, and `ForeachIdx3<i, j, k>` approach is that the compilation of hwy/tests/swizzle_test.cc is slow and consumes too much memory.

I had refactored the default implementation of `Per4LaneBlockShuffle<kIdx3, kIdx2, kIdx1, kIdx0>(v)` (which does table lookup) to detail::TblLookupPer4LaneBlkShuf, which takes a non-template `idx3210` parameter to allow detail::TblLookupPer4LaneBlkShuf to be unit tested using a loop.

I had also added a TestTblLookupPer4LaneBlkShuf function in TestPer4LaneBlockShuffle that checks the results of detail::TblLookupPer4LaneBlkShuf for all 256 cases using a loop.

Here is the loop that is used to check all 256 cases of detail::TblLookupPer4LaneBlkShuf:
```
    for (size_t idx3210 = 0; idx3210 <= 0xFF; idx3210++) {
      const size_t idx3 = (idx3210 >> 6) & 3;
      const size_t idx2 = (idx3210 >> 4) & 3;
      const size_t idx1 = (idx3210 >> 2) & 3;
      const size_t idx0 = idx3210 & 3;

      const auto actual = detail::TblLookupPer4LaneBlkShuf(v, idx3210);
      DoCheckPer4LaneBlkShufResult(d, N, actual, src_lanes, expected, idx3,
                                   idx2, idx1, idx0);
    }
```